NVIDIA-Merlin/Transformers4Rec
# [WIP] Define getting-started example notebook using TensorFlow & Add support of graph-mode

Fixes #283 #262 #258 #236

### Goals :soccer: and Implementation Details :construction:

TL;DR: This PR was initially created to fix the saving of TF model needed by Triton serving  and the definition of Tensorflow version of getting started notebook.  The issue had a larger scope than the saving method and I had to refactor the custom classes `MaskSequence`, `NextItemPredictionTask`, `RankingMetric` and `TransformerBlock`. The main changes are: 

- [x] Update HuggingFace dependency to 4.11* to support Albert and ELECTRA. (Note that ELECTRA model still need feature #263 for working correctly)

- [x] Re-factor the tf ops used to define the custom classes to make sure they are working in both modes: `eager` and `graph`. 

- [x] Define `get_config` / `from_config` methods to the custom classes for model saving and loading. 

- [x] Add a non trainable tf.Variable to `RankingMetric` classes as placeholder to store the updated states at each iteration. This is needed in `graph` mode as the metrics are part of the Keras class `NextItemPredictionTask`.

- [x] Improve the signature HF transformers `call` to fix the issue of shape mismatch when saving the model. This error is linked to non-used arguments being initialized with wrong tensors: The previous call method provided transformer_kwargs twice, e.g. `mask_perm` was set equal to `input_embed` instead of None).  

- [x] Extract from the HF model the serializable layer TF*MainLayer. In fact, all the HF models are defined with one custom Keras layer containing the architecture graph. (e.g. [tfXLNetModel](https://github.com/huggingface/transformers/blob/9e4ea2517504f144e74d9a356d58e9f2be32b3fa/src/transformers/models/xlnet/modeling_tf_xlnet.py#L1160))

- [x] Fix the loss function used by `NextItemPredictionTask`. We set `logits=true` in SparseCategoricalCrossentropy as the task already returns the logit scores. 

- [x] Expose compute_metrics flag of `compute_loss` methods to compute metrics during train / evaluation.  The method `process_metric` is added to flatten the list of scores returned by ranking metrics with respect to their threshold. 

- [x] Create a TensorFlow version of the getting started notebook, including model definition, training, evaluation, saving, and reloading.  

- [ ] Re-load the model for:  model serving, calling the method on a batch of data (`model()`), resume training (`model.fit()`), and compute evaluation scores on new data (`model.evaluate()`)  
  -  Using `.save()` and `tf.keras.models.load_model()` allow for saving the model's architecture and its weights. Applying the re-loaded model is working. However calling `fit()` and `evaluate()` methods raise an input signature error ```ValueError: Could not find matching function to call loaded from the SavedModel.```
  -  The second solution is to use weight saving and loading (`save_weights` / `load_weights`) but we need to construct the model before loading weight.  In this case, we can resume training without any error. 


### Testing Details :mag:


- Add tests to `test_model.py`: 
   - `test_simple_seq_classification`: Add a test of the sequence classification model and check if loss decreasing.
   - `test_next_item_fit`: Test, next item prediction task, using eager and graph modes.
   - `test_simple_model`: Test saving and loading of a simple classification model.
   - `test_save_load_item_prediction`: Test saving and loading of next item prediction task with a simple model using an `MLP` block.
   -  `test_save_load_transformer_model`: Test saving and loading of next item prediction task with supported HF transformers.
   - `test_resume_training`: Test resume training of next item prediction task using weight saving and loading

- Update `test_transformer.py` to include Albert config. 

- Add `test_score_different_from_zero` to `test_ranking_metrics.py` to check that computed metrics are different from 0.  


### Remaining tasks out of scope of this PR: 
- Fix scores of ranking metrics returned by the fit/evaluate methods in the getting-started notebook: The current scores are inconsistent for `mlm` (close to one).  In fact, the model is always returning the last non-padded item suggesting there is a leakage in the model definition (might be related to the masking schema computed by the `TabularSequenceFeatures` module). 

- For `clm`, reported RecallAt seems correct, while NDCGAt is returning the same score for all thresholds (the issue might be related to the nested usage of DCGAt metric)
